bokehjs/src/lib/models/widgets/tables/data_table.ts

```diff
@@ -110,19 +110,22 @@ export class TableDataProvider implements DataProvider<Item> {
   }
 
   sort(columns: SortColumn<Item>[]): void {
-    let cols = columns.map((column) => [column.sortCol.field, column.sortAsc ? 1 : -1] as const)
+    let cols = columns.map((column) => [column.sortCol, column.sortAsc ? 1 : -1] as const)
 
     if (cols.length == 0) {
-      cols = [[DTINDEX_NAME, 1]]
+      cols = [[{field: DTINDEX_NAME}, 1]]
     }
 
     const records = this.getRecords()
     const old_index = this.index.slice()
 
     this.index.sort((i0, i1) => {
-      for (const [field, sign] of cols) {
-        const v0 = records[old_index.indexOf(i0)][field!]
-        const v1 = records[old_index.indexOf(i1)][field!]
+      for (const [col, sign] of cols) {
+        const v0 = records[old_index.indexOf(i0)][col.field!]
+        const v1 = records[old_index.indexOf(i1)][col.field!]
+        if ((col as any).sorter != null) {
+          return sign * (col as any).sorter.compute(v0, v1)
+        }
```

Comment:
At minimum I would cast to `ColumnType`, which should give you access to `sorter`. Perhaps this could be done in `columns.map(...)` above.